src/cryptonote_core/loki_name_system.cpp

```diff
@@ -579,84 +608,163 @@ bool validate_lns_value_binary(uint16_t type, std::string const &value, std::str
+          *reason = err_stream.str();
+        }
+        return false;
       }
-      return false;
-    }
 
-    lns::owner_record owner = lns_db.get_owner_by_id(mapping.owner_id);
-    if (!owner)
-    {
-      if (reason)
+      uint64_t renew_window              = 0;
+      uint64_t expiry_blocks             = lns::lokinet_expiry_blocks(lns_db.network_type(), &renew_window);
+      uint64_t const renew_window_offset = expiry_blocks - renew_window;
+      uint64_t const min_renew_height    = mapping.register_height + renew_window_offset;
+
+      if (min_renew_height >= blockchain_height)
       {
-        error_stream_append_tx_msg(err_stream, tx, data) << "unexpected owner_id=" << mapping.owner_id << " does not exist";
+        print_tx_and_extra(err_stream, tx, data) << ", trying to renew too early, the earliest renew height=" << min_renew_height << ", urrent height=" << blockchain_height;
         *reason = err_stream.str();
+        return false; // Trying to renew too early
       }
-      return false;
-    }
 
-    if (requester.id != owner.id)
-    {
-      if (reason)
+      if (mapping.active(lns_db.network_type(), blockchain_height))
       {
-        error_stream_append_tx_msg(err_stream, tx, data) << " actual owner=" << owner.key << ", with owner_id=" << mapping.owner_id << ", does not match requester=" << requester.key << ", with id=" << requester.id;
-        *reason = err_stream.str();
+        // Lokinet entry expired i.e. it's no longer active. A purchase for this name is valid
+        // Check that the request originates from the owner of this mapping
+        lns::owner_record const requester = lns_db.get_owner_by_key(data.owner);
+        if (!requester)
+        {
+          if (reason)
+          {
+            print_tx_and_extra(err_stream, tx, data) << ", trying to renew existing mapping but owner specified in LNS extra does not exist, rejected";
+            *reason = err_stream.str();
+          }
+          return false;
+        }
+
+        lns::owner_record const owner = lns_db.get_owner_by_id(mapping.owner_id);
+        if (!owner)
+        {
+          if (reason)
+          {
+            print_tx_and_extra(err_stream, tx, data) << ", unexpected owner_id=" << mapping.owner_id << " does not exist";
+            *reason = err_stream.str();
+          }
+          return false;
+        }
+
+        if (requester.id != owner.id)
+        {
+          if (reason)
+          {
+            print_tx_and_extra(err_stream, tx, data) << ", actual owner=" << owner.key << ", with owner_id=" << mapping.owner_id << ", does not match requester=" << requester.key << ", with id=" << requester.id;
+            *reason = err_stream.str();
+          }
+          return false;
+        }
+      }
+      else
+      {
+        // NOTE: Mapping has expired, new purchase is valid
       }
```

Comment:
The `else {}` just to contain a comment seems rather superfluous.
`// else mapping has expired, new purchase is valid`
is one line instead of 4.

Reply:
Trimmed